tokio-trace-proc-macros/src/lib.rs

```diff
@@ -49,19 +32,20 @@ pub fn trace(_args: TokenStream, item: TokenStream) -> TokenStream {
         .collect();
     let param_names_clone = param_names.clone();
 
-    quote_spanned!(call_site=>
-        #(#attrs) *
-        #vis #constness #unsafety #asyncness #abi fn #ident(#params) #return_type {
-            span!(
-                tokio_trace::Level::TRACE,
-                #ident_str,
-                traced_function = &#ident_str
-                #(, #param_names = tokio_trace::field::debug(&#param_names_clone)),*
-            )
-            .enter(move || {
-                #block
-            })
-        }
-    )
-    .into()
+    let span = quote_spanned!(call_site =>
+        span!(
+            tokio_trace::Level::TRACE,
+            #ident_str,
+            traced_function = &#ident_str
+            #(, #param_names = tokio_trace::field::debug(&#param_names_clone)),*
+        )
+    );
+
+    function.block.stmts = quote_spanned!(call_site =>
+        #span.enter(move || {
+            #body
+        })
+    );
```

Comment:
This isn't blocking this PR, but I _think_ you might be able to write something like:

```rust
let is_async = function.asyncness.is_some();
function.block.stmts = if is_async {
    quote_spanned!(call_site =>
        #span.enter(move || {
            #body
        })
    )
} else {
    quote_spanned!(call_site =>
        #span.enter(move || {
            async move {
                #body
            }
        })
    )
}
```

...and get async functions be instrumented correctly.